src/predictor.rs

```diff
@@ -0,0 +1,753 @@
+                16 => rev_predict_f16(in_buf, out_buf, predictor_info.samples_per_pixel as _),
+                32 => rev_predict_f32(in_buf, out_buf, predictor_info.samples_per_pixel as _),
+                64 => rev_predict_f64(in_buf, out_buf, predictor_info.samples_per_pixel as _),
+                _ => {
+                    return Err(AsyncTiffError::General(format!(
+                        "thou shalt not predict f{bit_depth:?}"
+                    )))
+                }
+            }
+        }
+    } else {
+        // specially handle padding bytes
+        // create a buffer for the full width
+        let mut input = BytesMut::from(buffer);
+
+        let input_row_stride =
+            predictor_info.chunk_width as usize * predictor_info.bits_per_sample as usize / 8;
+        for (in_buf, out_buf) in input
+            .chunks_mut(input_row_stride)
+            .zip(res.chunks_mut(output_row_stride))
+        {
+            let mut out_row = BytesMut::zeroed(input_row_stride);
+            match bit_depth {
+                16 => rev_predict_f16(in_buf, &mut out_row, predictor_info.samples_per_pixel as _),
+                32 => rev_predict_f32(in_buf, &mut out_row, predictor_info.samples_per_pixel as _),
+                64 => rev_predict_f64(in_buf, &mut out_row, predictor_info.samples_per_pixel as _),
+                _ => {
+                    return Err(AsyncTiffError::General(format!(
+                        "thou shalt not predict f{bit_depth:?}"
+                    )))
+                }
+            }
+            // remove the padding bytes
+            out_buf.copy_from_slice(&out_row[..output_row_stride]);
+        }
+    }
+    Ok(res.into())
+}
+
+/// Reverse floating point prediction
+///
+/// floating point prediction first shuffles the bytes and then uses horizontal
+/// differencing
+/// also performs byte-order conversion if needed.
+pub fn rev_predict_f16(input: &mut [u8], output: &mut [u8], samples: usize) {
+    // reverse horizontal differencing
+    for i in samples..input.len() {
+        input[i] = input[i].wrapping_add(input[i - samples]);
+    }
+    // reverse byte shuffle and fix endianness
+    for (i, chunk) in output.chunks_mut(2).enumerate() {
+        chunk.copy_from_slice(&u16::to_ne_bytes(
+            // convert to native-endian
+            // floating predictor is be-like
+            u16::from_be_bytes([input[i], input[input.len() / 2 + i]]),
+        ));
+    }
+}
+
+/// Reverse floating point prediction
+///
+/// floating point prediction first shuffles the bytes and then uses horizontal
+/// differencing
+/// also performs byte-order conversion if needed.
+pub fn rev_predict_f32(input: &mut [u8], output: &mut [u8], samples: usize) {
+    // reverse horizontal differencing
+    for i in samples..input.len() {
+        input[i] = input[i].wrapping_add(input[i - samples]);
+    }
+    // reverse byte shuffle and fix endianness
+    for (i, chunk) in output.chunks_mut(4).enumerate() {
```

Comment:
Would [`chunks_exact_mut`](https://doc.rust-lang.org/std/primitive.slice.html#method.chunks_exact_mut) be ok to use here? Is there a chance that chunks may not have exactly 4 bytes at the end?

Reply:
should be ok, since the chunk-ness is based on `f32` being 4 bytes, also at the `f64` then